src/random_unitary.jl

```diff
@@ -0,0 +1,69 @@
+# Version of `sign` that returns one
+# if `x == 0`.
+function nonzero_sign(x)
+  iszero(x) && return one(x)
+  return sign(x)
+end
+
+using LinearAlgebra: LinearAlgebra, Diagonal, diag
+function qr_positive(M::AbstractMatrix)
+  Q, R = LinearAlgebra.qr(M)
+  Q′ = typeof(R)(Q)
+  signs = nonzero_sign.(diag(R))
+  Q′ = Q′ * Diagonal(signs)
+  R = Diagonal(conj.(signs)) * R
+  return Q′, R
+end
+
+using Random: Random, AbstractRNG
+
+dual(x) = x
+
+function random_unitary(
```

Comment:
I think it might be nice to define this similar to how `rand` and `rand!` work, where the former simply allocates something and the latter actually does the work.

Reply:
I agree that is a nice code pattern to follow, but because this is a tensor operation, you would have to fuse the input out-of-place anyway. Maybe that pattern would make more sense for the matrix version.

Reply:
FYI I'm trying this out now, including a way to rewrite it where do don't explicitly need `dual` in the library.